source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,693 @@
+        if (analysis::Const* c = FindRecordedConst(id)) {
+          if (IsValidTypeForComponentWiseOperation(c->type())) {
+            operands.push_back(c);
+            return true;
+          }
+        }
+        return false;
+      })) {
+    return nullptr;
+  }
+
+  if (result_type->AsInteger() || result_type->AsBool()) {
+    // Scalar operation
+    uint32_t result_val = OperateScalars(spec_opcode, operands);
+    auto result_const = CreateConst(result_type, {result_val});
+    return BuildInstructionAndAddToModule(std::move(result_const), pos);
+  } else {
+    // Vector operation
+    const analysis::Type* element_type =
+        result_type->AsVector()->element_type();
+    uint32_t num_dims = result_type->AsVector()->element_count();
+    std::vector<uint32_t> result_vec =
+        OperateVectors(spec_opcode, num_dims, operands);
+    std::vector<const analysis::Const*> result_vector_components;
+    for (uint32_t r : result_vec) {
+      auto rc = CreateConst(element_type, {r});
+      result_vector_components.push_back(rc.get());
+      BuildInstructionAndAddToModule(std::move(rc), pos);
+    }
+    auto new_vec_const = MakeUnique<analysis::VectorConst>(
+        result_type->AsVector(), result_vector_components);
+    return BuildInstructionAndAddToModule(std::move(new_vec_const), pos);
+  }
+}
+
+ir::Instruction*
+FoldSpecConstantOpAndCompositePass::BuildInstructionAndAddToModule(
+    std::unique_ptr<analysis::Const> c, ir::Module::inst_iterator* pos) {
+  analysis::Const* new_const = c.get();
+  uint32_t new_id = ++max_id_;
+  const_val_to_id_[new_const] = new_id;
+  id_to_const_val_[new_id] = std::move(c);
+  auto new_inst = CreateInstruction(new_id, new_const);
+  if (!new_inst) return nullptr;
+  auto* new_inst_ptr = new_inst.get();
+  *pos = pos->InsertBefore(std::move(new_inst));
+  (*pos)++;
+  def_use_mgr_->AnalyzeInstDefUse(new_inst_ptr);
+  return new_inst_ptr;
+}
+
+std::unique_ptr<analysis::Const>
+FoldSpecConstantOpAndCompositePass::CreateConstFromInst(ir::Instruction* inst) {
+  std::vector<uint32_t> literal_words_or_ids;
+  std::unique_ptr<analysis::Const> new_const;
+  // Collect the constant defining literals or component ids.
+  for (uint32_t i = 0; i < inst->NumInOperands(); i++) {
+    literal_words_or_ids.insert(literal_words_or_ids.end(),
+                                inst->GetInOperand(i).words.begin(),
+                                inst->GetInOperand(i).words.end());
+  }
+  switch (inst->opcode()) {
+    // OpConstant{True|Flase} have the value embedded in the opcode. So they
+    // are not handled by the for-loop above. Here we add the value explicitly.
+    case SpvOp::SpvOpConstantTrue:
+      literal_words_or_ids.push_back(true);
+      break;
+    case SpvOp::SpvOpConstantFalse:
+      literal_words_or_ids.push_back(false);
+      break;
+    case SpvOp::SpvOpConstantNull:
+    case SpvOp::SpvOpConstant:
+    case SpvOp::SpvOpConstantComposite:
+    case SpvOp::SpvOpSpecConstantComposite:
```

Comment:
Won't this incorrectly convert an OpSpecConstantComposite to an OpConstantComposite when you have a nested OpSpecConstantComposite that *can't* be converted?
E.g. something like:
  struct Pair { int a, b;} 
  struct Rect { Pair upperleft; Pair lowerright; }
If I had %r as OpSpecConstantComposite with elements %up %low and %up is not convertible, then this code will convert %r.
??